### ISSUE TITLE
snd-hda-intel "Too many BDL entries"

### ISSUE THREAD

**Issue by plbossart** [thesofproject/linux#4322]
Not really an SOF issue but something I've seen on my development NUC in the last weeks with the 5.19 Ubuntu kernel

````
[    5.315806] snd_hda_intel 0000:00:1f.3: enabling device (0000 -> 0002)
[    5.783807] snd_hda_intel 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
[    5.945428] snd_hda_intel 0000:00:1f.3: device 8086:2064 is on the power_save denylist, forcing power_save to 0
[    7.144302] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[   45.156458] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.480798] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.481494] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.482471] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.483159] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.485179] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.487878] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.488508] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.489391] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.490012] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.491624] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.494647] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.495432] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.496430] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.497211] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.498965] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.502179] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.503022] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.504080] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.504899] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.506618] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.508790] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.509267] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.509877] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.510307] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.511375] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.513212] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.513565] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.514032] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.514361] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.515213] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.517131] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.517490] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.517937] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.518277] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.519048] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.520976] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.521389] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.521798] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.522115] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.522794] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.547872] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.548086] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.548365] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.548566] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.549128] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.550644] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.550857] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.551135] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.551338] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.551861] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.553510] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.553776] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.554105] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.554356] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.554938] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[42167.556599] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.556877] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.557244] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[42167.557524] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[42167.558127] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[87596.994671] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[87596.994877] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[87596.995141] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[87597.010717] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[87597.010924] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[87597.011258] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[87597.011451] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[87597.011973] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=4194304, period=131072
[87597.013422] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[87597.034976] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
[87597.035255] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=1048576, period=32768
[87597.035634] snd_hda_intel 0000:00:1f.3: Too many BDL entries: buffer=2097152, period=65536
````

**Comment by fredoh9** [comment]
I tested with ADL-P RVP HDA, legacy hda mode. Playback and sleep 5 and capture in a loop. I don't see this error.
20230421 daily build, 6.3.0 rc3, 57d1c63077e3

**Comment by plbossart** [comment]
I think this was fixed by @tiwai, my development device does not track the latest kernel.

**Comment by keqiaozhang** [comment]
@plbossart , can we close this bug?

**Comment by Vamshigopal** [comment]
> I think this was fixed by @tiwai, my development device does not track the latest kernel.

@plbossart  We see the same issue in one of the  CML based chrome designs when we moved to 5.15 kernel from 4.19 kernel. 
Can you please share the fix , so we can test and backport the fix if it works for us.

**Comment by plbossart** [comment]
Try one of the changes in sound/core/memalloc.c


db918321275d ALSA: memalloc: don't use GFP_COMP for non-coherent dma allocations
3306877aff07 ALSA: memalloc: don't pass bogus GFP_ flags to dma_alloc_*
cc2651637406 ALSA: memalloc: Allocate more contiguous pages for fallback case
9d8e536d36e7 (tag: sound-fix-6.1-rc5) ALSA: memalloc: Try dma_alloc_noncontiguous() at first
9736a325137b ALSA: memalloc: Don't fall back for SG-buffer with IOMMU
a61c7d88d38c ALSA: memalloc: use __GFP_RETRY_MAYFAIL for DMA mem allocs
37137ec26c2c ALSA: hda: Once again fix regression of page allocations with IOMMU
dd164fbfdc20 ALSA: memalloc: Drop special handling of GFP for CONTINUOUS allocation
